pkg/issuer/acme/dns/webhook/webhook.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2018 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"github.com/golang/glog"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/jetstack/cert-manager/pkg/issuer/acme/dns/util"
+)
+
+const (
+	maxBodyLength = 5 * 1024 * 1024
+
+	presentOperation operation = "present"
+	cleanupOperation operation = "cleanup"
+)
+
+type operation string
+
+type WebhookPayload struct {
+	Operation operation `json:"operation"`
```

Comment:
Use of an unexported var here makes it impossible for anyone to construct a WebhookPayload from outside of this package, making it difficult for those implementing a webhook to do so (as they cannot access the different operation types).

Could we perhaps move the actual API types for this into a new package, `pkg/issuer/acme/dns/webhook/api`? In time I think this should probably move to be a part of our regular API types too (i.e. in `pkg/apis/...`) - but we don't need to rush to do this now IMO 😄 (let's iterate instead!)